0015-how-to-document-cnd-http-api.adoc

```diff
@@ -0,0 +1,207 @@
+swaps` and `GET /
+swaps/rfc003/:id`. JSON Schema is supported by OpenAPI, https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.1.md#schema-object[with some caveats]. There are https://apisyouwonthate.com/blog/solving-openapi-and-json-schema-divergence[ways to get around this situation], and full support for JSON Schema is https://github.com/OAI/OpenAPI-Specification/pull/1977[in the works].
+
+=== How to generate the specification ===
+
+==== Options ====
+
+===== Automatically =====
+* Switch to actix-web and use https://paperclip.waffles.space/actix-plugin.html[this experimental plugin].
+* Wait for Rocket to https://github.com/SergioBenitez/Rocket/issues/297[implement this feature] and switch back to it.
+* https://github.com/seanmonstar/warp/issues/89[Wait for this feature] to come to warp.
+
+===== Manually =====
+* Use https://github.com/swagger-api/swagger-editor[Swagger Editor] (over 5k stars on GitHub).
+* Use a plugin for an editor/IDE (look at the editors listed https://openapi.tools/[here]).
+
+===== Assisted =====
+* Using https://inspector.swagger.io/builder[Swagger Inspector].
+* Make API calls to a running cnd through the UI and generate part of the specification.
+
+==== Recommendation ====
+Unfortunately the tools aren't there yet to automatically produce the API specification from source code in Rust. There seems to be some interest for this, so it may come in the future. For the time being, we will have to write it ourselves. Fortunately, this specification will be custom-id:test-specification[tested] and we can use tools such as https://speccy.io/[speccy] to validate it.
+
+=== How to visualise specification ===
+
+==== Options ====
+
+===== https://petstore.swagger.io/[SwaggerUI] =====
+* Extremely popular (over 15k stars on https://github.com/swagger-api/swagger-ui[GitHub])
+* RPC-style.
+* Looks like this:
+image::How_to_visualise_documentation/Swagger_UI_2019-08-29_18-16-23.png[scaledwidth=100%]
+
+* Hides JSON bodies.
+
+===== https://redocly.github.io/redoc/[ReDoc] =====
+* Popular (over 1k stars on GitHub).
+* Very https://github.com/Redocly/redoc/#deployment[simple] to set up.
+* Three-column style.
+* It looks very good:
+image::How_to_visualise_documentation/redoc-demo_2019-08-29_18-29-34.png[scaledwidth=100%]
+
+===== http://cheesestore.github.io/[Spectacle] =====
+* Quite popular (about 1k stars on GitHub).
+* Looks similar to ReDoc:
+image::How_to_visualise_documentation/screenshot_2019-08-29_18-46-24.jpg[scaledwidth=100%]
+
+==== Recommendation ====
+ReDoc because it is easy and I think it looks nice.
+
+=== What does the documentation include? ===
+
+==== HTTP API Specification. ====
+Models services.
+
+==== JSON Schema. ====
+Models data.
+
+=== How to host documentation for current `master` HEAD ===
+If we go for the recommended option of using custom-id:redoc[ReDoc], it's https://www.npmjs.com/package/redoc#tldr[very easy]:
```

Comment:
what is this `custom-id:redoc[ReDoc]` ?

Reply:
Broken link due to automatic conversion from org to asciidoc. Fixed now.